### PR TITLE
feat: support advanced initialization in v5

### LINKED ISSUE #9455: Advanced initialization with Auth.Js 5
### What is the improvement or update you wish to see?

Missing documentation on advanced initialization for the latest version of Next Auth.

### Is there any context that might help us understand?

I am trying to specify session max age based on request data. I need to set max to one day session if user is authenticating using only password auth, but if user is having account then longer. Now as per documentation I need to use advanced initialization, but as per latest version of Next Auth, it is not possible to pass req and res to NextAuth.
How to achieve this with latest version?

### Does the docs page already exist? Please link to it.

https://authjs.dev/guides/basics/initialization#advanced-initialization

### PR DESCRIPTION
**Pull request by ThangHuuVu**
## ☕️ Reasoning

Add support for advanced initialization for v5. 

## 🧢 Checklist

- [ ] Documentation
- [ ] Tests
- [ ] Ready to be merged

## 🎫 Affected issues

Fixes: #9455 

## 📌 Resources

- [Security guidelines](https://github.com/nextauthjs/.github/blob/main/SECURITY.md)
- [Contributing guidelines](https://github.com/nextauthjs/.github/blob/main/CONTRIBUTING.md)
- [Code of conduct](https://github.com/nextauthjs/.github/blob/main/CODE_OF_CONDUCT.md)
- [Contributing to Open Source](https://kcd.im/pull-request)

### CODE REVIEW

File: packages/next-auth/src/index.tsx [hunk 1]
```diff
@@ -45,6 +45,26 @@
...
+ *  if (req) {
+ *   console.log(req) // do something with the request
+ *   return { providers: [ GitHub ] }
+ *  }
+ *  return { providers: [ GitHub ] }
+ * })
+ * ```
+ *
+ * :::tip
+ * This is useful if you want to customize the configuration based on the request, for example, to add a different provider based on the request URL's `searchParams`.
```
Comment (balazsorban44) [inline packages/next-auth/src/index.tsx:65]:
```suggestion
 * This is useful if you want to customize the configuration based on the request, for example, to add a different provider in staging/dev environments.
```

File: packages/next-auth/src/index.tsx [hunk 2]
```diff
@@ -61,8 +81,12 @@ import type {
   NextApiResponse,
 } from "next"
 import type { AppRouteHandlerFn } from "next/dist/server/future/route-modules/app-route/module.js"
-import type { NextRequest } from "next/server"
-import type { NextAuthConfig, NextAuthRequest } from "./lib/index.js"
+import { NextRequest } from "next/server"
```
Comment (balazsorban44) [inline packages/next-auth/src/index.tsx:84]:
```suggestion
import type { NextRequest } from "next/server"
```

**Review by balazsorban44 — CHANGES_REQUESTED** [review]

**Review by balazsorban44 — APPROVED** [review]

**Comment by chrisivo** [comment]
Is there any way lazy initialisation can be executed asynchronously? I need to make async calls to obtain the Id's and secrets for the providers. I could do this using v4, but seems like there's no support in v5 beta (so far)?

**Comment by ndom91** [comment]
@chrisivo what do you mean exactly? The first argument to `NextAuth()` can be an async function. So yuo can do this (as is shown in most examples):

```ts
export const { auth, signIn, signOut } = NextAuth({
  providers: [ ... ],
  debug: true
})
```

But yuo can also do this:

```ts
export const { auth, signIn, signOut } = NextAuth(async (req) => {
  return {
    providers: [ ... ],
    debug: true
  }
))
```

**Comment by chrisivo** [comment]
Thanks for the response ndom91. Is there any roadmap at all for introducing asynchronous functions when initialising in `next-auth` ?

**Comment by maiconcarraro** [comment]
Facing same situation from @chrisivo, need to generate secret for apple sign in during initialization :(

**Comment by nicolaspernoud** [comment]
Hello, 
Same here, we need to get provider configuration from prisma (async) ...

**Comment by ThangHuuVu** [comment]
@chrisivo @maiconcarraro @nicolaspernoud I opened a new PR to add support for asynchronous lazy init - thanks for flagging the issue

**Pull request merged**